File: q0198/q0198_iterative.java
class Solution {
    public int rob(int[] nums) {
        // base cases
        int n = nums.length;
        if (n == 1)
            return nums[0];
        else if (n == 2)
            return Math.max(nums[0], nums[1]);

        // Let dp[i] represents the max return until the i-th house
        int[] dp = new int[n];
        dp[0] = nums[0];
        dp[1] = Math.max(nums[0], nums[1]);

        for (int i = 2; i < n; i++) {
            int prevPrevHouse = nums[i] + dp[i - 2];
            int prevHouse = dp[i - 1];
            dp[i] = Math.max(prevPrevHouse, prevHouse);
        }

        // Final return is dp[n-1] as the robber will consider until the last house
        return dp[n - 1];
    }
}

File: q0198/q0198_iterative.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]
        elif n == 2:
            return max(nums[0], nums[1])
        dp = [0] * n
        dp[0] = nums[0]
        dp[1] = max(nums[0], nums[1])
        for i in range(2, n):
            res1 = nums[i] + dp[i - 2]
            res2 = dp[i - 1]
            dp[i] = max(res1, res2)
        return dp[-1]